WordPress/src/main/java/org/wordpress/android/ui/uploads/AutoSavePostIfNotDraftUseCase.kt

```diff
@@ -0,0 +1,108 @@
+package org.wordpress.android.ui.uploads
+
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.suspendCancellableCoroutine
+import org.greenrobot.eventbus.Subscribe
+import org.greenrobot.eventbus.ThreadMode.BACKGROUND
+import org.wordpress.android.fluxc.Dispatcher
+import org.wordpress.android.fluxc.generated.PostActionBuilder
+import org.wordpress.android.fluxc.model.CauseOfOnPostChanged.RemoteAutoSavePost
+import org.wordpress.android.fluxc.model.LocalOrRemoteId.LocalId
+import org.wordpress.android.fluxc.model.PostModel
+import org.wordpress.android.fluxc.store.PostStore
+import org.wordpress.android.fluxc.store.PostStore.OnPostChanged
+import org.wordpress.android.fluxc.store.PostStore.OnPostStatusFetched
+import org.wordpress.android.fluxc.store.PostStore.RemotePostPayload
+import org.wordpress.android.ui.uploads.AutoSavePostIfNotDraftResult.PostAutoSaved
+import org.wordpress.android.ui.uploads.AutoSavePostIfNotDraftResult.PostIsDraftInRemote
+import javax.inject.Inject
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.resume
+
+// TODO: Not be so verbose in naming
+
+private const val DRAFT_POST_STATUS = "draft"
+
+interface OnAutoSavePostIfNotDraftCallback {
+    fun handleAutoSavePostIfNotDraftResult(result: AutoSavePostIfNotDraftResult)
+}
+
+sealed class AutoSavePostIfNotDraftResult {
+    data class PostIsDraftInRemote(val post: PostModel) : AutoSavePostIfNotDraftResult()
+    data class PostAutoSaved(val post: PostModel) : AutoSavePostIfNotDraftResult()
+    // TODO: Add error result types
+}
+
+// TODO: Add documentation
+// TODO: Add unit tests
+class AutoSavePostIfNotDraftUseCase @Inject constructor(
+    private val dispatcher: Dispatcher,
+    private val postStore: PostStore
+) {
+    private val postStatusContinuations = HashMap<LocalId, Continuation<String>>()
+    private val autoSaveContinuations = HashMap<LocalId, Continuation<PostModel>>()
+
+    init {
+        dispatcher.register(this)
+    }
+
+    fun autoSavePostOrUpdateDraft(
+        remotePostPayload: RemotePostPayload,
+        callback: OnAutoSavePostIfNotDraftCallback
+    ) {
+        val localPostId = LocalId(remotePostPayload.post.id)
+        if (postStatusContinuations.containsKey(localPostId) ||
+                autoSaveContinuations.containsKey(localPostId)) {
+            // TODO: Consider canceling the previous job instead
```

Comment:
Is that because we are going to use this from `PostUploadHandler`? If that's the case, I don't like a reusable component to have to rely on it's caller's implementation and I avoid it as much as I can.

Reply:
I was more thinking about throwing "IllegalStateException" with explanation that it shouldn't be invoked with the same id until the first request completes - this would work with any client. Wdyt?

Reply:
Sure, we can do that.